lib/connection.js

```diff
@@ -341,14 +341,16 @@ Connection.prototype.ready = function(n){
  * @api private
  */
 
-Connection.prototype.finish = function(id){
+Connection.prototype.finish = function(id, fn){
   assertValidMessageId(id);
-  var fn = this.onerror;
   var self = this;
-  if (!this._ready) return fn(new Error('cannot finish, connection not ready'));
+  if (!this._ready) return this.onerror(new Error('cannot finish, connection not ready'));
   this.command('FIN', [id], function(){
     self.emit('finish', id);
     --self.inFlight;
+    if (fn) {
+      fn(id, self.inFlight);
```

Comment:
Why are we passing `self.inFlight` to the callback? I assume the caller has access to the reader and can check inflight count?